### ISSUE TITLE
Web Components Driver

### ISSUE THREAD

**Issue by TylorS** [cyclejs/cycle-core#173]
I was wondering if there is any interest in some form of a web components driver. The DOM driver is likely to remove custom elements in the future so maybe having another solution could be useful. I haven't thought any of it out just yet but was hoping to see what/if you guys would be interested in.

I'm currently wondering if creating some sort of streamlined way to create the following in Cycle would be beneficial to some of us

1.) Vanilla Web Components (Custom Elements, Html Imports, template tags, shadow DOM)
3.) Mozilla's [X-Tag](http://www.x-tags.org/)
3.) Google's [Polymer]
4.) Something else, please give details.

Please cast your votes!

**Comment by HighOnDrive** [comment]
Cycle is a runtime that is well suited to run components of all stripes. I think anything goes, however I'd want to check if the component types support a MVI structure. Might be best to wrap or simply call them from inside a proper Cycle component, likely in View and then let the Intent handle event processing.

So, it would be nice and possible to support them but my vote is just to treat them as injectables, as essentially dummy plugins. The other thing to check would be Virtual-DOM compatibility, do all these alternate types integrate well with the Virtual-DOM?

**Comment by Frikki** [comment]
5

**Comment by staltz** [comment]
"Web Components" and "Driver" don't chime that well together in my mind. A Driver is an FRP interface for imperative commands. Web Components are a DOM interface for a packaged component.

That said, there are two things we could do with WCs in Cycle.js:

1 - Make easy helpers to use WCs inside Cycle DOM.
2 - Make easy helpers to package a Cycle DOM application (`main`) as a WC

Actually both can exist, they aren't mutually exclusive.

1 - I know WCs can be easily used in virtual-dom, but the importing process may not be that smooth yet in a Cycle.js workflow. Maybe it would make sense to build a small proof of concept project to make sure it is already quite easy to use WCs in a Cycle app. If it's not easy, then we will discover what's missing.

2 - This one I believe doesn't yet exist and would require some helpers to be built. By the way you should check https://github.com/erykpiast/cyclejs-wc which was made specifically for this, however for an already older version of Cycle.

**Comment by TylorS** [comment]
Yes, now actually thinking about it Driver was the wrong term to use,
should have been "utility" or something similar.

Both options seem very interesting, but I think I am looking more towards
option 1.

I know that web components may or may not have a future, and I also know
the power of nested dialogues are a very compelling, but it surprises me
how many people are still trying to use custom elements. Which led me to
wonder what people would think about using web components with Cycle in an
easy to process way, with some kind of library to make it more idiomatic.

**Comment by staltz** [comment]
Yes, I really agree.

**Comment by gregpardo** [comment]
I'd like to see a repository of pure cycle components that have well documented input sources and output sinks. I was thinking of some sort of spec via comments or yaml that can be searched for and parsed into a database and added to a public API that would return metadata about the component/s (screenshots, readme, inputs, outputs). Basically you could submit your url to your repository and it would search and parse out the components within/poll and update them. Does this idea make any sense? I'm looking for a pet Elixir project might be a good fit if it's at all useful.

**Comment by staltz** [comment]
Sounds great @gregpardo, please go ahead if you have the energy.

**Comment by micahscopes** [comment]
I'm very interested in **#1**.  I'm new to cycle.js, but composable web components is one of my primary interests in using cycle.js in the first place.  Right now, my workflow is to slowly piece together various libraries, each with a different level of functional-ness, into a new web component.  This is usually pretty messy, especially as dealing with web components can be messy already.   Learning cycle.js seems like an excellent way to make modular component development more comprehensible.
